blackbird_python/blackbird/program.py

```diff
@@ -0,0 +1,290 @@
+and return an instance of the :class:`BlackbirdProgram` class.
+
+Summary
+-------
+
+.. autosummary::
+    numpy_to_blackbird
+    BlackbirdProgram
+
+Code details
+~~~~~~~~~~~~
+"""
+import numpy as np
+
+
+def numpy_to_blackbird(A, var_name):
+    """Converts a numpy array to a Blackbird script array type.
+
+    Args:
+        A (array): 2-dimensional NumPy array
+        var_name (str): the array variable name
+
+    Returns:
+        list[str]: list containing each line representing the
+            Blackbird array variable declaration
+    """
+
+    if np.any(np.iscomplex(A)):
+        # complex array
+        script = ["complex array {}[{}, {}] =".format(var_name, *A.shape)]
+        for row in A:
+            row_str = "    " + ", ".join(
+                ["{0}{1}{2}j".format(n.real, "+-"[int(n.imag < 0)], abs(n.imag)) for n in row]
+            )
+            script.append(row_str)
+
+    elif np.all(np.mod(A, 1) == 0):
+        # integer array
+        script = ["int array {}[{}, {}] =".format(var_name, *A.shape)]
+        for row in A:
+            row_str = "    " + ", ".join(["{}".format(int(n)) for n in row])
+            script.append(row_str)
+
+    elif not np.any(np.iscomplex(A)):
+        # real array
+        script = ["float array {}[{}, {}] =".format(var_name, *A.shape)]
+        for row in A:
+            row_str = "    " + ", ".join(["{}".format(n) for n in row])
+            script.append(row_str)
+
+    script.append("")
+
+    return script
+
+
+class BlackbirdProgram:
+    """Python representation of a Blackbird program."""
+
+    def __init__(self, name="blackbird_program", version=1.0):
+        self._var = {}
+        self._modes = set()
+
+        # the following attributes fully describe a Blackbird program
+        self._name = name
+        self._version = version
+        self._target = {"name": None, "options": dict()}
+        self._operations = []
+
+    @property
+    def name(self):
+        """Name of the Blackbird program
+
+        Returns:
+            str: name
+        """
+        return self._name
+
+    @property
+    def version(self):
+        """Version of the Blackbird parser the program targets
+
+        Returns:
+            float: version number
+        """
+        return self._version
+
+    @property
+    def modes(self):
+        """A set of non-negative integers specifying the mode numbers the program manipulates.
+
+        Returns:
+            set[int]: mode numbers
+        """
+        return self._modes
+
+    @property
+    def target(self):
+        """Contains information regarding the target device of the quantum
+        program (i.e., the target device the Blackbird script is compiled for).
+
+        Important keys include:
+
+        * ``'name'`` (str): the name of the device the Blackbird script requests to be
+            run on. If no target is requested, the returned value will be ``None``.
```

Comment:
just checking: return value will be `None` string?

Reply:
return type will just be `NoneType`. I'll make this clearer. Just easier to check `if name is None` as opposed to `if name == 'None'`, which can lead to inconsistencies.